tokio/src/util/sharded_list.rs

```diff
@@ -0,0 +1,160 @@
+#![cfg_attr(not(feature = "full"), allow(dead_code))]
+
+use std::{ptr::NonNull, sync::atomic::Ordering};
+
+use crate::loom::sync::{Mutex, MutexGuard};
+use std::sync::atomic::AtomicUsize;
+
+use super::linked_list::{Link, LinkedList};
+
+/// An intrusive  linked list supporting high concurrent updates.
+///
+/// It currently relies on `LinkedList`, so it is the caller's
+/// responsibility to ensure the list is empty before dropping it.
+///
+/// Note: Due to its inner sharded design, the order of node cannot be guaranteed.
+pub(crate) struct ShardedList<L, T> {
+    lists: Box<[Mutex<LinkedList<L, T>>]>,
+    count: AtomicUsize,
+    shard_mask: usize,
+}
+
+/// Defines the id of a node in a linked list, different ids cause inner list nodes
+/// to be scattered in different mutexs.
+///
+/// # Safety
+///
+/// Implementations must guarantee that `Target` types are pinned in memory. In
+/// other words, when a node is inserted, the value will not be moved as long as
+/// it is stored in the list.
+pub(crate) unsafe trait ShardedListItem: Link {
+    // The returned id is used to pick which list this item should go into.
+    unsafe fn get_shared_id(target: NonNull<Self::Target>) -> usize;
+}
+
+impl<L, T> ShardedList<L, T> {
+    /// Creates a new and empty sharded linked list under specified size
+    pub(crate) fn new(sharded_size: usize) -> Self {
+        // Find power-of-two sizes best matching arguments
+        let mut size = 1;
+        while size < sharded_size {
+            size <<= 1;
+        }
```

Comment:
I'm not a big fan of this. It means that the size is not always accurate, and the loop will run infinitely if `sharded_size > 2^63`.
```suggestion
        assert!(sharded_size.is_power_of_two());
```

Reply:
I also think it's better to leave the size of `sharded_size` to  the caller.

But considering that the number of worker threads is not necessarily exactly an integer power of 2, this loop must still be executed.

However, spawn_concurrency_level is 4 times the number of worker threads. I considered setting its maximum value to 1024, which I believe is a large enough number.

Reply:
I moved this logic to 
```
impl Builder {
    fn spawn_concurrency_level(&self) -> usize {
        const MAX_SPAWN_CONCURRENCY_LEVEL: usize = 1 << 16;

        match self.spawn_concurrency_level {
            Some(i) => i,
            None => {
                use crate::loom::sys::num_cpus;
                let core_threads = self.worker_threads.unwrap_or_else(num_cpus);

                let mut size = 1;
                while size / 4 < core_threads && size < MAX_SPAWN_CONCURRENCY_LEVEL {
                    size <<= 1;
                }
                size.min(MAX_SPAWN_CONCURRENCY_LEVEL)
            }
        }
    }
}
```